### ISSUE TITLE
[BUG - IMPORTANT] Data loss: XP and levels reset for some users

### ISSUE THREAD

**Issue by LoneDev6** [mcMMO-Dev/mcMMO#4158]
3 users reported that their XP and levels got reset.
I know they are not lying because I checked my backups and mcmmo.users file had high levels and xp for these users.
I stopped the server, copied and pasted the old file, started again and everything was fine and restored.
After a restart their levels and XP got reset again.

Anyone else found that issue after updating to the latest version of the plugin? 
mcmmo: 2.1.126-SNAPSHOT
server: paper 1.15.2 #183

**Comment by NIronwolf** [comment]
A similar thing has happened on my server.  My host uses a shared disk space and when it's full I get a zero'd out data file.  Might I suggest that saves happen to a temporary file first (to ensure data integrity) then move that file to the real data file name?

**Comment by LoneDev6** [comment]
I'm glad someone else confirms that issue, but I don't think that was because of full disk as it wasn't full, other plugins were working fine with no disk full errors.

I think mcMMO needs to implement something different than plain files, a sqlite database would be better for sure

**Comment by Draycia** [comment]
https://github.com/mcMMO-Dev/mcMMO/blob/master/src/main/resources/config.yml#L144-L170 ?

**Comment by LoneDev6** [comment]
> 
> 
> https://github.com/mcMMO-Dev/mcMMO/blob/master/src/main/resources/config.yml#L144-L170 ?

I don't need this, I said "sqlite" not MySQL

**Comment by Draycia** [comment]
No, you said `something different than plain text files`, you gave sqlite as an example.

**Comment by LoneDev6** [comment]
Anyway, this is not the problem. The issue is that levels and XP for some users got reset randomly.
A solution would be to automate some sort of backup each day and allow admins to use a command like /mmorestore <backup_name> <player>

this is the easiest solution and will be very handly in case of problems

**Comment by NIronwolf** [comment]
So there is a backup directory next to the flatfile directory that looks like it's taking snapshots fairly frequently.  There just isn't an in game command to manage the data.  I don't know how I missed it before but I think it will help both of us.  Just a bit of manual work to get the data out.

**Comment by LoneDev6** [comment]
> 
> 
> So there is a backup directory next to the flatfile directory that looks like it's taking snapshots fairly frequently. There just isn't an in game command to manage the data. I don't know how I missed it before but I think it will help both of us. Just a bit of manual work to get the data out.

A command like this will be really useful to avoid having to manually edit the file and risk breaking it:
`/mmorestore <backup_name> <player_name>`

**Comment by LoneDev6** [comment]
> 
> 
> have you check your log files ? i got same problem caused by reaching the Mojang profile API limit during UUID Fetch. the reason behind the xp and level reset happen because mcmmo requesting more than 600 profile in less than 10 minutes and because of that i got temporary blocked from Mojang API, which made some UUID unable to be fetch properly and their

So this is a bug, when server cannot fetch online UUID mcmmo uses the offline UUID and this would result in data reset? Right?

**Comment by aerotracker** [comment]
> > have you check your log files ? i got same problem caused by reaching the Mojang profile API limit during UUID Fetch. the reason behind the xp and level reset happen because mcmmo requesting more than 600 profile in less than 10 minutes and because of that i got temporary blocked from Mojang API, which made some UUID unable to be fetch properly and their
> 
> So this is a bug, when server cannot fetch online UUID mcmmo uses the offline UUID and this would result in data reset? Right?

i wont called it a bug, imo the plugin itself would fetch UUID flawlessly if mojang have higher numbers on their limit, and probably yes, since there are no way to confirm the online UUID after the blockage from the API, so mcmmo use the offline UUID. so as long my player still in the first batch, their data basically safe. the solution that i use is by slowing the fetching process and when i say slow it is really slow, since my server have 2000+ UUID, the fetching process took 30 minutes to make sure we dont get any temp block from Mojang API

**Comment by LoneDev6** [comment]
A solution could be to cache the UUIDs in a file as they won't change
during time, this would avoid getting rate limited from Mojang servers

2020-06-12 15:54 GMT+02:00, Aerotracker <notifications@github.com>:
